File: Backend/Product.cpp
#include "Product.h"
#include <sstream>

namespace std {

Product::Product(Database* db) {
    this->db = db;
}

string Product::getAllProducts() {
    PGresult* res = db->executeQuery("SELECT * FROM Products");
    stringstream ss;
    int rows = PQntuples(res);
    for (int i = 0; i < rows; ++i) {
        ss << "Product ID: " << PQgetvalue(res, i, 0)
           << ", Seller ID: " << PQgetvalue(res, i, 1)
           << ", Name: " << PQgetvalue(res, i, 2)
           << ", Description: " << PQgetvalue(res, i, 3)
           << ", Price: $" << PQgetvalue(res, i, 4)
           << ", Category: " << PQgetvalue(res, i, 5)
           << ", Image URL: " << PQgetvalue(res, i, 6) << "\n";
    }
    PQclear(res);
    return ss.str();
}

string Product::getProductByID(int productID) {
    string query = "SELECT * FROM Products WHERE productID = " + to_string(productID);
    PGresult* res = db->executeQuery(query);
    stringstream ss;
    if (PQntuples(res) > 0) {
        ss << "Product ID: " << PQgetvalue(res, 0, 0)
           << ", Seller ID: " << PQgetvalue(res, 0, 1)
           << ", Name: " << PQgetvalue(res, 0, 2)
           << ", Description: " << PQgetvalue(res, 0, 3)
           << ", Price: $" << PQgetvalue(res, 0, 4)
           << ", Category: " << PQgetvalue(res, 0, 5)
           << ", Image URL: " << PQgetvalue(res, 0, 6);
    } else {
        ss << "Product not found.";
    }
    PQclear(res);
    return ss.str();
}

}

File: Frontend(sql.js version)/Javascript/product.js

function initProductDb(SQL) {
    let pdb;
    const saved = localStorage.getItem("productDatabase");
    if (saved) {
      pdb = new SQL.Database(Uint8Array.from(JSON.parse(saved)));
    } else {
      pdb = new SQL.Database();
      pdb.run(`
        CREATE TABLE IF NOT EXISTS products (
          productID   INTEGER PRIMARY KEY,
          name        TEXT,
          description TEXT,
          price       REAL,
          owner       INTEGER,
          imageURL    TEXT
        );
      `);
      localStorage.setItem(
        "productDatabase",
        JSON.stringify(Array.from(pdb.export()))
      );
    }
    return pdb;
  }
  
  (async () => {
    // 1) Load the SQL.js runtime
    const SQL = await initSqlJs({ locateFile: f => `Javascript/${f}` });
  
    // 2) Initialize or load the products DB
    const db = initProductDb(SQL);
  
    // 3) Read ?id= from the URL
    const params = new URLSearchParams(window.location.search);
    const id     = params.get("id");
    if (!id) {
      document.getElementById("prod-name").textContent = "Product not found";
      return;
    }
  
    // 4) Query that product
    const result = db.exec(`
      SELECT name, price, description, imageURL
      FROM products
      WHERE productID = ${id}
    `)[0];
  
    // 5) If empty, show error
    if (!result) {
      document.getElementById("prod-name").textContent = "Product not found";
      return;
    }
  
    // 6) Destructure and render
    const [name, price, description, imageURL] = result.values[0];
    document.getElementById("prod-name").textContent        = name;
    document.getElementById("prod-price").textContent       = `$${price}`;
    document.getElementById("prod-description").textContent = description;
    if (imageURL) {
      document.getElementById("prod-image").src = imageURL;
    }
  
    // 7) Wire up Add to Cart
    document.getElementById("add-to-cart")
      .addEventListener("click", () => {
        const cart = JSON.parse(localStorage.getItem("cart") || "[]");
        const existing = cart.find(item => item.id === id);
        if (existing) {
          existing.qty++;
        } else {
          cart.push({ id, name, price: Number(price), qty: 1 });
        }
        localStorage.setItem("cart", JSON.stringify(cart));
        alert(`✅ Added "${name}" to your cart!`);
      });
  })();
  